src/main/scala/org/apache/hadoop/fs/sftpwithseek/SFTPFileSystem.scala

```diff
@@ -0,0 +1,593 @@
+        throw new IOException(e)
+    }
+    val absoluteSrc = makeAbsolute(workDir, src)
+    val absoluteDst = makeAbsolute(workDir, dst)
+    if (!exists(channel, absoluteSrc))
+      throw new IOException(String.format(SFTPFileSystem.E_SPATH_NOTEXIST, src))
+    if (exists(channel, absoluteDst))
+      throw new IOException(String.format(SFTPFileSystem.E_DPATH_EXIST, dst))
+    var renamed = true
+    try {
+      channel.cd("/")
+      channel.rename(src.toUri.getPath, dst.toUri.getPath)
+    } catch {
+      case e: SftpException =>
+        renamed = false
+    }
+    renamed
+  }
+  @throws[IOException]
+  override def initialize(uriInfo: URI, conf: Configuration): Unit = {
+    super.initialize(uriInfo, conf)
+    setConfigurationFromURI(uriInfo, conf)
+    setConf(conf)
+    this.uri = uriInfo
+  }
+  override def getScheme   = "sftp"
+  override def getUri: URI = uri
+  @throws[IOException]
+  override def open(f: Path, bufferSize: Int): FSDataInputStream = {
+    val channel       = connect
+    var workDir: Path = null
+    try workDir = new Path(channel.pwd)
+    catch {
+      case e: SftpException =>
+        throw new IOException(e)
+    }
+    var absolute = makeAbsolute(workDir, f)
+    val fileStat = getFileStatus(channel, absolute)
+    if (fileStat.isDirectory) {
+      disconnect(channel)
+      throw new IOException(String.format(SFTPFileSystem.E_PATH_DIR, f))
+    }
+    var is: InputStream = null
+    try { // the path could be a symbolic link, so get the real path
+      absolute = new Path("/", channel.realpath(absolute.toUri.getPath))
+      is = channel.get(absolute.toUri.getPath)
+    } catch {
+      case e: SftpException =>
+        throw new IOException(e)
+    }
+    val fis = new FSDataInputStream(new SFTPInputStream(is, channel, statistics))
+    fis
+  }
+
+  /**
+   * A stream obtained via this call must be closed before using other APIs of
+   * this class or else the invocation will block.
+   */ @throws[IOException]
+  override def create(
+      f: Path,
+      permission: FsPermission,
+      overwrite: Boolean,
+      bufferSize: Int,
+      replication: Short,
+      blockSize: Long,
+      progress: Progressable
+  ): FSDataOutputStream = {
+    val client        = connect
+    var workDir: Path = null
+    try workDir = new Path(client.pwd)
+    catch {
+      case e: SftpException =>
+        throw new IOException(e)
+    }
+    val absolute = makeAbsolute(workDir, f)
+    if (exists(client, f))
+      if (overwrite) delete(client, f, false)
+      else {
+        disconnect(client)
+        throw new IOException(String.format(SFTPFileSystem.E_FILE_EXIST, f))
+      }
+    var parent = absolute.getParent
+    if (parent == null || !mkdirs(client, parent, FsPermission.getDefault)) {
+      parent = if (parent == null) new Path("/") else parent
+      disconnect(client)
+      throw new IOException(String.format(SFTPFileSystem.E_CREATE_DIR, parent))
+    }
+    var os: OutputStream = null
```

Comment:
Maybe I’m out of context, but why are we adding this awful code in here?

Reply:
It is explained in the PR description. 

SFTP file system support in Hadoop is outdated, so several projects have updated it independently, we have done the same.

It is a copy of org.apache.hadoop.fs.sftp classes with the fixes applied.

Reply:
I see all the comments are about the style of these classes. 
They are an external implementation, and I would prefer to address any change in the SFTP support classes, if wanted, in another issue.

Reply:
Moved the classes to their original Java implemention in: 25d34a0e31b595893e2f9bd3fe392ba9591ef249